### PR TITLE
configury: define C11 macros once

### PR DESCRIPTION
**Pull request by ggouaillardet**
Revamp OPAL_PROG_CC_C11 macro in order to define macros only once.
Otherwise, macros get redefined during the configure process and
issue a bunch of warning in config.log

Signed-off-by: Gilles Gouaillardet <gilles@rist.or.jp>

### CODE REVIEW

**Comment by hjelmn** [comment]
Looks good to me.

File: config/opal_setup_cc.m4 [hunk 1]
```diff
@@ -47,30 +41,42 @@ AC_DEFUN([OPAL_PROG_CC_C11_HELPER],[
...
+
+           AC_DEFINE_UNQUOTED([OPAL_C_HAVE__ATOMIC], [$opal_prog_cc_c11_helper__Atomic_available],
+                              [Whether C compiler supports __Atomic keyword])
+
+           AC_DEFINE_UNQUOTED([OPAL_C_HAVE__GENERIC], [$opal_prog_cc_c11_helper__Generic_available],
+                              [Whether C compiler supports __Generic keyword])
+
+           AC_DEFINE_UNQUOTED([OPAL_C_HAVE__STATIC_ASSERT], [$opal_prog_cc_c11_helper__static_assert_available],
+                              [Whether C compiler support _Static_assert keyword])])
+
```
Comment (jsquyres) [inline config/opal_setup_cc.m4]: I'm not sure I understand this block -- don't we *always* want to AC_DEFINE this values (and make sure that they are 0 or 1)?

Comment (hjelmn) [reply config/opal_setup_cc.m4]: Hmm, indeed. We do. I haven't tested this with a non-C11 compiler. @ggouaillardet Might be worth doing. Anything older than gcc 4.9.0 will work. I have 4.8.x installed specifically to test with.

Comment (ggouaillardet) [reply config/opal_setup_cc.m4]: i run a CentOS 7 box with gcc 4.8.x, which is a non-C11 compiler.

In this environment and on `master`, `OPAL_PROG_CC_C11_HELPER` is called several times in `OPAL_PROG_CC_C11` (test the C compiler without flag and then with all flags since all these test fail). Finally, `OPAL_PROG_CC_C11_HELPER` is invoked a last time in `OPAL_SETUP_CC` since `$opal_cv_c11_supported = no`.

With a C11 compiler, then `OPAL_PROG_CC_C11_HELPER` is invoked until it success.

I did not change this behavior in this PR.
In order to __always__ `AC_DEFINE` the macros, I added a forth argument to force the macros to be `AC_DEFINE`'ed, and this argument is only `yes` in `OPAL_SETUP_CC` since we know at this stage we do not have a C11 compiler.

Bottom line, I think this PR is correct.
And that being said, we might want to take this opportunity to revisit the existing logic.
Feel free to suggest anything and I will be happy to update this PR.

**Comment by benmenadue** [comment]
@ggouaillardet Okay, thanks, it's autogen-ing at the moment. I also just pushed my own similar change into my fork ([benmenadue:rework-c11-detection](/benmenadue/ompi/tree/rework-c11-detection)), which only tests for the features once it works out if and how C11 is supported. Couldn't PR it because I forgot to sign off the commits. But if your PR works, then it's all good 👍  .

**Comment by benmenadue** [comment]
@ggouaillardet `configure` finished fine, and I only got one copy of those macros :+1:

File: config/opal_setup_cc.m4 [hunk 2]
```diff
@@ -47,29 +45,42 @@ AC_DEFUN([OPAL_PROG_CC_C11_HELPER],[
...
+                              [Whether C compiler support atomic convenience variables in stdatomic.h])
+
+           AC_DEFINE_UNQUOTED([OPAL_C_HAVE__ATOMIC], [$opal_prog_cc_c11_helper__Atomic_available],
+                              [Whether C compiler supports __Atomic keyword])
+
+           AC_DEFINE_UNQUOTED([OPAL_C_HAVE__GENERIC], [$opal_prog_cc_c11_helper__Generic_available],
+                              [Whether C compiler supports __Generic keyword])
+
+           AC_DEFINE_UNQUOTED([OPAL_C_HAVE__STATIC_ASSERT], [$opal_prog_cc_c11_helper__static_assert_available],
+                              [Whether C compiler support _Static_assert keyword])])
```
Comment (jsquyres) [inline config/opal_setup_cc.m4]: Why not move these AC_DEFINEs down into OPAL_SETUP_CC?  That way, you don't need a 4th argument and a bogus extra call into OPAL_PROG_CC_C11_HELPER.

**Comment by ggouaillardet** [comment]
@jsquyres I made the changes (and I should have done this in the first place ...)

can you please double check them ?

File: config/opal_setup_cc.m4 [hunk 3]
```diff
@@ -157,9 +151,26 @@ AC_DEFUN([OPAL_SETUP_CC],[
 
         # Get the correct result for C11 support flags now that the compiler flags have
         # changed
-        OPAL_PROG_CC_C11_HELPER([],[],[])
+        OPAL_PROG_CC_C11_HELPER([], [], [])
```
Comment (jsquyres) [inline config/opal_setup_cc.m4]: Do you know what this is for?  We only get here if we have already determined that there is no C11 support, so we check for C99 support.  ...but then we check for C11 support again, with no CLI options.

Are we just checking to see if the C99 flags also enable *some* C11 functionality?  If so, the comment above this line is misleading, at best.

Comment (ggouaillardet) [reply config/opal_setup_cc.m4]: My guess is we want to make sure all C11 macros are set to zero.

@hjelmn can you please comment on that ?

**Comment by jsquyres** [comment]
bot:ompi:retest

File: config/opal_setup_cc.m4 (hunk 3 above)
Comment (hjelmn) [reply config/opal_setup_cc.m4]: That check is there because the C99 support check may add additional flags (for example ```--std=gnu99```) which may change the results. This is entirely defensive. Example (have not seen this happen), without a flag: _Generic = 1, with ```--std=gn99``` _Generic = 0.

Comment (hjelmn) [reply config/opal_setup_cc.m4]: It may never happen but that was the original reason. The line dates to the earliest versions of the C11 checks.

**Comment by thananon** [comment]
Seems like I broke mergability with #5272 . Sorry.

**Comment by ggouaillardet** [comment]
No problem, I will fix that later today

**Comment by ibm-ompi** [comment]
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/23b302387a2a091a4836f4f60b699dce

**Comment by ibm-ompi** [comment]
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ea85c619ac6e84e4940b2d77f0de89c8

**Comment by ibm-ompi** [comment]
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/c1e83ed663612112f4f1437f4725f5d2

**Comment by ibm-ompi** [comment]
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ac59e4dadff1a4c942a828a7ac55696a

**Comment by ibm-ompi** [comment]
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/804bd915ea7a9ca464fb717ee68e8aca
